src/coreclr/vm/classlayoutinfo.cpp

```diff
@@ -533,20 +534,45 @@ namespace
                 }
     #endif
                 MetaSig fsig(pCOMSignature, cbCOMSignature, pModule, pTypeContext, MetaSig::sigField);
-                CorElementType corElemType = fsig.NextArgNormalized();
+                CorElementType corElemType = fsig.NextArg();
+
                 TypeHandle typeHandleMaybe;
                 if (corElemType == ELEMENT_TYPE_VALUETYPE) // Only look up the next element in the signature if it is a value type to avoid causing recursive type loads in valid scenarios.
                 {
-                    typeHandleMaybe = fsig.GetLastTypeHandleThrowing(ClassLoader::LoadTypes,
-                        CLASS_LOAD_APPROXPARENTS,
-                        TRUE);
+                    SigPointer::HandleRecursiveGenericsForFieldLayoutLoad recursiveControl;
+                    recursiveControl.pModuleWithTokenToAvoidIfPossible = pModule;
+                    recursiveControl.tkTypeDefToAvoidIfPossible = cl;
+                    typeHandleMaybe = fsig.GetArgProps().GetTypeHandleThrowing(pModule,
+                                                                             pTypeContext,
+                                                                             ClassLoader::LoadTypes,
+                                                                             CLASS_LOAD_APPROXPARENTS,
+                                                                             TRUE, NULL, NULL, NULL, 
+                                                                             &recursiveControl);
+
+                    if (typeHandleMaybe.IsNull())
+                    {
+                        // Everett C++ compiler can generate a TypeRef with RS=0
+                        // without respective TypeDef for unmanaged valuetypes,
+                        // referenced only by pointers to them.
+                        // In such case, GetTypeHandleThrowing returns null handle,
+                        // and we return E_T_VOID
```

Comment:
I think we should update this.

Reply:
Probably, but this change is confusing enough as it is without adding a known potential behavior change. I'd like to see that done as a separate PR.

Reply:
If you do that, please get rid of the weirdo ThrowButNullV11McppWorkaround logic too.